README.md
# fastrand

[![Build](https://github.com/smol-rs/fastrand/workflows/Build%20and%20test/badge.svg)](
https://github.com/smol-rs/fastrand/actions)
[![License](https://img.shields.io/badge/license-Apache--2.0_OR_MIT-blue.svg)](
https://github.com/smol-rs/fastrand)
[![Cargo](https://img.shields.io/crates/v/fastrand.svg)](
https://crates.io/crates/fastrand)
[![Documentation](https://docs.rs/fastrand/badge.svg)](
https://docs.rs/fastrand)

Extension functionality for the [`fastrand`] crate.

This crate contains code that may be of some use to users of [`fastrand`]. Code contained in
this crate is not included in [`fastrand`] due to either the niche not being large enough to
justify the new functionality or for semver concerns.

## Usage

Various functions are exposed in this crate as top-level functions. These manipulate the global
thread-local RNG and can be used without any local state. Note that these require the `"std"`
default feature to be enabled.

```
use fastrand_contrib::f32_range;

let x = f32_range(1.5..3.0);
assert!(x >= 1.5 && x < 3.0);
```

To extend [`fastrand::Rng`], import the [`RngExt`] trait.

```
use fastrand_contrib::RngExt;
```

Now, all new methods are available on [`fastrand::Rng`].

```
use fastrand::Rng;
use fastrand_contrib::RngExt;

let mut rng = Rng::with_seed(0x1234);
let x = rng.f32_range(1.5..3.0);
assert!(x >= 1.5 && x < 3.0);
```

[`fastrand`]: https://crates.io/crates/fastrand
[`fastrand::Rng`]: https://docs.rs/fastrand/latest/fastrand/struct.Rng.html

# Features

- `std` (enabled by default): Enables the `std` library. Freestanding functions only work with this feature enabled. Also enables the `fastrand/std` feature.
- `libm`: Uses [`libm`](https://crates.io/crates/libm) dependency for math functions in `no_std` environment.

## License

Licensed under either of

 * Apache License, Version 2.0 ([LICENSE-APACHE](LICENSE-APACHE) or http://www.apache.org/licenses/LICENSE-2.0)
 * MIT license ([LICENSE-MIT](LICENSE-MIT) or http://opensource.org/licenses/MIT)

at your option.

#### Contribution

Unless you explicitly state otherwise, any contribution intentionally submitted
for inclusion in the work by you, as defined in the Apache-2.0 license, shall be
dual licensed as above, without any additional terms or conditions.


CHANGELOG.md
# Version 0.1.0

- Initial release



benches/normal.rs
#![feature(test)]

extern crate test;

use fastrand::Rng;
use fastrand_contrib::RngExt;
use test::Bencher;

const SEED: u64 = 42;
const MU: f64 = 10.0;
const SIGMA: f64 = 3.0;

#[bench]
fn box_muller(b: &mut Bencher) {
    let mut rng = Rng::with_seed(SEED);

    b.iter(|| {
        let mu = core::hint::black_box(MU);
        let sigma = core::hint::black_box(SIGMA);

        // https://en.wikipedia.org/wiki/Box%E2%80%93Muller_transform
        let u1 = loop {
            let u1 = rng.f64();

            if u1 > f64::EPSILON {
                break u1;
            }
        };

        let u2 = rng.f64();
        let mag = sigma * (-2.0 * u1.ln()).sqrt();
        let output = mag * (core::f64::consts::TAU * u2).cos() + mu;

        core::hint::black_box(output);
    });
}

#[bench]
fn standard_approximation(b: &mut Bencher) {
    let mut rng = Rng::with_seed(SEED);

    b.iter(|| {
        let mu = core::hint::black_box(MU);
        let sigma = core::hint::black_box(SIGMA);

        // http://marc-b-reynolds.github.io/distribution/2021/03/18/CheapGaussianApprox.html
        let u = rng.u128(..);

        let mask = 0xffffffff;
        let a = (u & mask) as i64;
        let b = ((u >> 32) & mask) as i64;
        let c = ((u >> 64) & mask) as i64;
        let d = (u >> 96) as i64;

        // Magic constant.
        let k = 3.97815e-10;

        let output = k * ((a + b) - (c + d)) as f64 * sigma + mu;

        core::hint::black_box(output);
    });
}

#[bench]
fn popcount_approximation(b: &mut Bencher) {
    let mut rng = Rng::with_seed(SEED);

    b.iter(|| {
        let mu = core::hint::black_box(MU);
        let sigma = core::hint::black_box(SIGMA);

        // http://marc-b-reynolds.github.io/distribution/2021/03/18/CheapGaussianApprox.html
        let u = rng.u128(..);

        let bd = (u << 64).count_ones() as i64 - 32;

        let a = ((u >> 64) & 0xffffffff) as i64;
        let b = (u >> 96) as i64;

        let td = a - b;

        let r = ((bd << 32) + td) as f64;

        // Magic constant.
        let k = 5.76917e-11;

        let output = k * r * sigma + mu;

        core::hint::black_box(output);
    })
}

#[bench]
fn f64_normal(b: &mut Bencher) {
    let mut rng = Rng::with_seed(SEED);

    b.iter(|| {
        let mu = core::hint::black_box(MU);
        let sigma = core::hint::black_box(SIGMA);

        let output = rng.f64_normal(mu, sigma);

        core::hint::black_box(output);
    });
}

#[bench]
fn f64_normal_approx(b: &mut Bencher) {
    let mut rng = Rng::with_seed(SEED);

    b.iter(|| {
        let mu = core::hint::black_box(MU);
        let sigma = core::hint::black_box(SIGMA);

        let output = rng.f64_normal_approx(mu, sigma);

        core::hint::black_box(output);
    });
}


src/float_range.rs
use core::ops::{Add, Bound, Div, Mul, Neg, RangeBounds, Sub};

use crate::BaseRng;

pub(super) fn f32(rng: &mut impl BaseRng, range: impl RangeBounds<f32>) -> f32 {
    float_range_impl(rng, range)
}

pub(super) fn f64(rng: &mut impl BaseRng, range: impl RangeBounds<f64>) -> f64 {
    float_range_impl(rng, range)
}

trait FloatExt:
    Add<Self, Output = Self>
    + Sub<Self, Output = Self>
    + Mul<Self, Output = Self>
    + Div<Self, Output = Self>
    + Neg<Output = Self>
    + Copy
    + Sized
{
    const MIN: Self;
    const MAX: Self;
    const HALF: Self;

    fn is_finite(self) -> bool;
    /// Generate a random float in [0, 1) range.
    fn gen_close_01_open(rng: &mut impl BaseRng) -> Self;
    /// Generate a random float in (0, 1] range.
    fn gen_open_01_close(rng: &mut impl BaseRng) -> Self;
    /// Generate a random float in (0, 1) range.
    fn gen_open_01_open(rng: &mut impl BaseRng) -> Self;
    /// Get the maximum float that can be generated in [0, 1) range (i.e., the
    /// float just below 1).
    fn max_rand() -> Self;
}

macro_rules! impl_float_ext {
    ($float:ident, $max_rand_bits:literal) => {
        impl FloatExt for $float {
            const MIN: Self = $float::MIN;
            const MAX: Self = $float::MAX;
            const HALF: Self = 0.5;

            #[inline]
            fn is_finite(self) -> bool {
                $float::is_finite(self)
            }
            #[inline]
            fn gen_close_01_open(rng: &mut impl BaseRng) -> Self {
                rng.$float()
            }
            #[inline]
            fn gen_open_01_close(rng: &mut impl BaseRng) -> Self {
                1.0 - rng.$float()
            }
            #[inline]
            fn gen_open_01_open(rng: &mut impl BaseRng) -> Self {
                loop {
                    let r = rng.$float();

                    if r != 0.0 {
                        return r;
                    }
                }
            }
            #[inline]
            fn max_rand() -> Self {
                <$float>::from_bits($max_rand_bits)
            }
        }
    };
}

// Max rand constant is conceptually 0.9999... in the precision of the
// particular float type. It can be determined in the following steps:
//
//   1. Fill the fractional part of the float with 1s (in binary).
//   2. Move the number to have exponent 0 (i.e., the exponent is equal to the
//      exponent bias).
//   3. Since the float representation uses implicit leading 1, subtract 1.0
//      from the number constructed in the previous steps.
//
// This is a code snippet for f32:
//
// ```
// let fraction_bits = 23; // Significand bits without the implicit leading 1.
// let exponent_bias = 127;
// let discard_bits = u32::BITS - fraction_bits;
// let max_rand = f32::from_bits(((u32::MAX >> discard_bits) | (exponent_bias << fraction_bits))) - 1.0;
// ```

impl_float_ext!(f32, 0x3f7ffffe);
impl_float_ext!(f64, 0x3feffffffffffffe);

/// Indication whether a range is exclusive, inclusive on either side or
/// inclusive on both sides. `Bound::Unbounded` is treated as inclusive, because
/// we use MIN and MAX constants for such bounds.
enum Inclusive {
    None,
    Left,
    Right,
    Both,
}

impl Inclusive {
    fn from_bounds<T>(range: impl RangeBounds<T>) -> Self {
        match (range.start_bound(), range.end_bound()) {
            (Bound::Excluded(_), Bound::Excluded(_)) => Inclusive::None,
            (Bound::Included(_), Bound::Excluded(_)) | (Bound::Unbounded, Bound::Excluded(_)) => {
                Inclusive::Left
            }
            (Bound::Excluded(_), Bound::Included(_)) | (Bound::Excluded(_), Bound::Unbounded) => {
                Inclusive::Right
            }
            (Bound::Included(_), Bound::Included(_))
            | (Bound::Included(_), Bound::Unbounded)
            | (Bound::Unbounded, Bound::Included(_))
            | (Bound::Unbounded, Bound::Unbounded) => Inclusive::Both,
        }
    }
}

fn float_range_impl<T: FloatExt>(rng: &mut impl BaseRng, range: impl RangeBounds<T>) -> T {
    let low = match range.start_bound() {
        Bound::Included(&low) | Bound::Excluded(&low) => low,
        Bound::Unbounded => T::MIN,
    };

    let high = match range.end_bound() {
        Bound::Included(&high) | Bound::Excluded(&high) => high,
        Bound::Unbounded => T::MAX,
    };

    let inclusive = Inclusive::from_bounds(range);

    // Our generator is able to generate floats with one or both sides of the
    // range open. However, it can't generate a float from the range closed on
    // both sides. For this case, we divide the scale by maximum random number
    // which "stretches" the range to include both sides. This is the approach
    // used in rand crate:
    // https://github.com/rust-random/rand/blob/f3dd0b885c4597b9617ca79987a0dd899ab29fcb/src/distributions/uniform.rs#L953
    let scale = match inclusive {
        Inclusive::None | Inclusive::Left | Inclusive::Right => high - low,
        Inclusive::Both => (high - low) / T::max_rand(),
    };

    if scale.is_finite() {
        // Generate a random number between 0 and 1, where the bounds are
        // included based on the desired range inclusiveness.
        let r = match inclusive {
            Inclusive::None => T::gen_open_01_open(rng),
            Inclusive::Right => T::gen_open_01_close(rng),
            Inclusive::Left => T::gen_close_01_open(rng),
            // Inclusiveness on both sides is achieved by stretching the scale
            // above.
            Inclusive::Both => T::gen_close_01_open(rng),
        };

        r * scale + low
    } else {
        // Scale not being finite means that the range is wider than the float
        // type can represent (or that at least one side is not finite). In such
        // case, we need to fall back into the following technique which does a
        // bit more work but can handle such ranges. Source:
        // https://medium.com/analytics-vidhya/random-floats-in-any-range-9b40d30b637b
        let high_half = T::HALF * high;
        let low_half = T::HALF * low;
        let mid_point = high_half + low_half;

        // Decide if we generate the value to the right or left from the middle
        // point. We always want to have a chance that the middle point is
        // sampled, so we can't use the (0, 1] trick with one-side inclusive
        // ranges. That is why we stretch those in appropriate cases.
        let (r, stretch) = if rng.bool() {
            let stretch = match inclusive {
                Inclusive::None | Inclusive::Left => false,
                Inclusive::Right | Inclusive::Both => true,
            };
            let r = T::gen_close_01_open(rng);
            (r, stretch)
        } else {
            let stretch = match inclusive {
                Inclusive::None | Inclusive::Right => false,
                Inclusive::Left | Inclusive::Both => true,
            };
            let r = -T::gen_close_01_open(rng);
            (r, stretch)
        };

        let half_scale = if stretch {
            let half_scale = (high_half - low_half) / T::max_rand();
            if half_scale.is_finite() {
                half_scale
            } else {
                // If the range is so extreme that it can't be stretched,
                // use the standard scale.
                high_half - low_half
            }
        } else {
            high_half - low_half
        };

        r * half_scale + mid_point
    }
}

#[cfg(test)]
mod tests {
    use fastrand::Rng;

    use super::*;

    #[test]
    fn f32_range_in_bounds() {
        let mut rng = Rng::with_seed(42);

        let range = -2.0..2.0;
        for _ in 0..10000 {
            assert!(range.contains(&float_range_impl(&mut rng, range.clone())));
        }
    }

    #[test]
    fn f32_range_wide_range_in_bounds() {
        let mut rng = Rng::with_seed(42);

        let range = f32::MIN..f32::MAX;
        for _ in 0..10000 {
            assert!(range.contains(&float_range_impl(&mut rng, range.clone())));
        }
    }

    #[test]
    fn f32_range_unbounded_finite() {
        let mut rng = Rng::with_seed(42);

        let range = ..;
        for _ in 0..10000 {
            assert!(&float_range_impl::<f32>(&mut rng, range).is_finite());
        }
    }
}


src/float_normal.rs
use core::{
    cmp::PartialOrd,
    ops::{Add, Mul, Neg},
};

use crate::BaseRng;

#[cfg(any(feature = "std", feature = "libm"))]
pub(super) fn f32(rng: &mut impl BaseRng, mu: f32, sigma: f32) -> f32 {
    float_normal_impl(rng, mu, sigma)
}

#[cfg(any(feature = "std", feature = "libm"))]
pub(super) fn f64(rng: &mut impl BaseRng, mu: f64, sigma: f64) -> f64 {
    float_normal_impl(rng, mu, sigma)
}

pub(super) fn f32_approx(rng: &mut impl BaseRng, mu: f32, sigma: f32) -> f32 {
    float_normal_approx_impl(rng, mu, sigma)
}

pub(super) fn f64_approx(rng: &mut impl BaseRng, mu: f64, sigma: f64) -> f64 {
    float_normal_approx_impl(rng, mu, sigma)
}

trait FloatExt:
    Add<Self, Output = Self> + Mul<Self, Output = Self> + Neg<Output = Self> + PartialOrd<Self> + Sized
{
    const EPSILON: Self;

    fn from_f64(x: f64) -> Self;
    fn gen(rng: &mut impl BaseRng) -> Self;
}

#[cfg(any(feature = "std", feature = "libm"))]
trait FloatMathExt: FloatExt {
    const TAU: Self;

    fn ln(self) -> Self;
    fn sqrt(self) -> Self;
    fn cos(self) -> Self;
}

macro_rules! impl_float_ext {
    ($float:ident) => {
        impl FloatExt for $float {
            const EPSILON: Self = $float::EPSILON;

            #[inline]
            fn from_f64(x: f64) -> Self {
                x as $float
            }
            #[inline]
            fn gen(rng: &mut impl BaseRng) -> Self {
                rng.$float()
            }
        }
    };
}

macro_rules! impl_float_math_ext {
    ($float:ident, $tau:ident) => {
        #[cfg(all(feature = "std", not(feature = "libm")))]
        impl FloatMathExt for $float {
            const TAU: Self = $tau;

            #[inline]
            fn ln(self) -> Self {
                $float::ln(self)
            }
            #[inline]
            fn sqrt(self) -> Self {
                $float::sqrt(self)
            }
            #[inline]
            fn cos(self) -> Self {
                $float::cos(self)
            }
        }

        #[cfg(feature = "libm")]
        impl FloatMathExt for $float {
            const TAU: Self = $tau;

            #[inline]
            fn ln(self) -> Self {
                libm_dep::Libm::<$float>::log(self)
            }
            #[inline]
            fn sqrt(self) -> Self {
                libm_dep::Libm::<$float>::sqrt(self)
            }
            #[inline]
            fn cos(self) -> Self {
                libm_dep::Libm::<$float>::cos(self)
            }
        }
    };
}

// TAU constant was stabilized in Rust 1.47. Our current MSRV is 1.43.
#[cfg(any(feature = "std", feature = "libm"))]
#[allow(clippy::excessive_precision)]
const F32_TAU: f32 = 6.28318530717958647692528676655900577_f32;
#[cfg(any(feature = "std", feature = "libm"))]
#[allow(clippy::excessive_precision)]
const F64_TAU: f64 = 6.28318530717958647692528676655900577_f64;

impl_float_ext!(f32);
impl_float_ext!(f64);
impl_float_math_ext!(f32, F32_TAU);
impl_float_math_ext!(f64, F64_TAU);

#[cfg(any(feature = "std", feature = "libm"))]
fn float_normal_impl<T: FloatMathExt>(rng: &mut impl BaseRng, mu: T, sigma: T) -> T {
    // https://en.wikipedia.org/wiki/Box%E2%80%93Muller_transform
    let u1 = loop {
        let u1 = T::gen(rng);

        if u1 > T::EPSILON {
            break u1;
        }
    };

    let u2 = T::gen(rng);
    let mag = sigma * (-T::from_f64(2.0) * u1.ln()).sqrt();
    mag * (T::TAU * u2).cos() + mu
}

fn float_normal_approx_impl<T: FloatExt>(rng: &mut impl BaseRng, mu: T, sigma: T) -> T {
    // http://marc-b-reynolds.github.io/distribution/2021/03/18/CheapGaussianApprox.html
    let u = rng.u128();

    // Counting ones in a u64 half of the generated number gives us binomial
    // distribution with p = 1/2 and n = 64. Subtracting 32 centers the
    // distribution on [-32, 32]. Shifting the lower u64 by 64 bits discard the
    // other half and `count_ones` can be used without any masking.
    let bd = (u << 64).count_ones() as i64 - 32;

    // Sample two u32 integers from uniform distribution.
    let a = ((u >> 64) & 0xffffffff) as i64;
    let b = (u >> 96) as i64;

    // First iteration of Central limit theorem (summing two uniform random
    // variables) _often_ gives triangular distribution. By using subtraction
    // instead of addition, the triangular distribution is centered around zero.
    let td = a - b;

    // Sum the binomial and triangular distributions.
    let r = ((bd << 32) + td) as f64;

    // Magic constant for scaling which is a result of minimizing the maximum
    // error with respect to the reference normal distribution.
    let k = 5.76917e-11;

    T::from_f64(k * r) * sigma + mu
}

#[cfg(test)]
mod tests {
    use fastrand::Rng;

    use super::*;

    fn normal_distribution_test<F>(sample: F)
    where
        F: Fn(&mut Rng, f32, f32) -> f32,
    {
        // The test based on the following picture from Wikipedia:
        // https://upload.wikimedia.org/wikipedia/commons/thumb/3/3a/Standard_deviation_diagram_micro.svg/1920px-Standard_deviation_diagram_micro.svg.png
        let mut rng = Rng::with_seed(42);

        let mu = 10.0;
        let sigma = 3.0;

        let total = 10000;
        let mut in_one_sigma_range = 0;
        let mut in_two_sigma_range = 0;
        let mut in_three_sigma_range = 0;
        for _ in 0..total {
            let value = sample(&mut rng, mu, sigma);

            if (mu - sigma..=mu + sigma).contains(&value) {
                in_one_sigma_range += 1;
            } else if (mu - sigma * 2.0..=mu + sigma * 2.0).contains(&value) {
                in_two_sigma_range += 1;
            } else if (mu - sigma * 3.0..=mu + sigma * 3.0).contains(&value) {
                in_three_sigma_range += 1;
            }
        }

        let in_one_sigma_range = in_one_sigma_range as f32 / total as f32 * 100.0;
        let in_two_sigma_range = in_two_sigma_range as f32 / total as f32 * 100.0;
        let in_three_sigma_range = in_three_sigma_range as f32 / total as f32 * 100.0;
        assert!(
            (64.0..=72.0).contains(&in_one_sigma_range),
            "value in \"one sigma range\" should be sampled ~68.2%, but is {}%",
            in_one_sigma_range
        );
        assert!(
            (23.0..=31.0).contains(&in_two_sigma_range),
            "value in \"two sigma range\" should be sampled ~27.2%, but is {}%",
            in_two_sigma_range
        );
        assert!(
            (1.0..=7.0).contains(&in_three_sigma_range),
            "value in \"three sigma range\" should be sampled ~4.2%, but is {}%",
            in_three_sigma_range
        );
    }

    #[test]
    #[cfg(any(feature = "std", feature = "libm"))]
    fn normal_is_actually_normal() {
        normal_distribution_test(float_normal_impl);
    }

    #[test]
    fn normal_approx_is_actually_normal() {
        normal_distribution_test(float_normal_approx_impl);
    }
}


src/lib.rs
//! Extension functionality for the [`fastrand`] crate.
//!
//! This crate contains code that may be of some use to users of [`fastrand`]. Code contained in
//! this crate is not included in [`fastrand`] due to either the niche not being large enough to
//! justify the new functionality or for semver concerns.
//!
//! ## Usage
//!
//! Various functions are exposed in this crate as top-level functions. These manipulate the global
//! thread-local RNG and can be used without any local state. Note that these require the `"std"`
//! default feature to be enabled.
//!
//! ```
//! # #[cfg(feature = "std")] {
//! use fastrand_contrib::f32_range;
//!
//! let x = f32_range(1.5..3.0);
//! assert!(x >= 1.5 && x < 3.0);
//! # }
//! ```
//!
//! To extend [`fastrand::Rng`], import the [`RngExt`] trait.
//!
//! ```
//! use fastrand_contrib::RngExt;
//! ```
//!
//! Now, all new methods are available on [`fastrand::Rng`].
//!
//! ```
//! use fastrand::Rng;
//! use fastrand_contrib::RngExt;
//!
//! let mut rng = Rng::with_seed(0x1234);
//! let x = rng.f32_range(1.5..3.0);
//! assert!(x >= 1.5 && x < 3.0);
//! ```
//! # Features
//!
//! - `std` (enabled by default): Enables the `std` library. Freestanding functions only work with this
//!   feature enabled. Also enables the `fastrand/std` feature.
//! - `libm`: Uses [`libm`] dependency for math functions in `no_std` environment.
//!
//! Note that some functions are not available in `no_std` context if `libm` feature is not enabled.
//!
//! [`fastrand`]: https://crates.io/crates/fastrand
//! [`fastrand::Rng`]: https://docs.rs/fastrand/latest/fastrand/struct.Rng.html
//! [`libm`]: https://crates.io/crates/libm

#![cfg_attr(not(feature = "std"), no_std)]
#![forbid(unsafe_code, future_incompatible, missing_docs)]
#![doc(
    html_favicon_url = "https://raw.githubusercontent.com/smol-rs/smol/master/assets/images/logo_fullsize_transparent.png"
)]
#![doc(
    html_logo_url = "https://raw.githubusercontent.com/smol-rs/smol/master/assets/images/logo_fullsize_transparent.png"
)]

mod float_normal;
mod float_range;

use core::ops::RangeBounds;

pub use fastrand::{self, Rng};

trait BaseRng {
    fn f32(&mut self) -> f32;
    fn f64(&mut self) -> f64;
    fn bool(&mut self) -> bool;
    fn u128(&mut self) -> u128;
}

impl BaseRng for Rng {
    #[inline]
    fn f32(&mut self) -> f32 {
        Rng::f32(self)
    }
    #[inline]
    fn f64(&mut self) -> f64 {
        Rng::f64(self)
    }
    #[inline]
    fn bool(&mut self) -> bool {
        Rng::bool(self)
    }
    #[inline]
    fn u128(&mut self) -> u128 {
        Rng::u128(self, ..)
    }
}

#[cfg(feature = "std")]
struct GlobalRng;

#[cfg(feature = "std")]
impl BaseRng for GlobalRng {
    #[inline]
    fn f32(&mut self) -> f32 {
        fastrand::f32()
    }
    #[inline]
    fn f64(&mut self) -> f64 {
        fastrand::f64()
    }
    #[inline]
    fn bool(&mut self) -> bool {
        fastrand::bool()
    }
    #[inline]
    fn u128(&mut self) -> u128 {
        fastrand::u128(..)
    }
}

macro_rules! define_ext {
    ($(
        $(#[$meta:meta])*
        fn $name:ident(&mut self, $($argname:ident:$argty:ty),*) -> $ret:ty => $imp:path;
    )*) => {
        /// Extra methods for [`fastrand::Rng`].
        pub trait RngExt {
            $(
            $(#[$meta])*
            fn $name(&mut self, $($argname: $argty),*) -> $ret;
            )*
        }

        impl RngExt for Rng {
            $(
            $(#[$meta])*
            fn $name(&mut self, $($argname: $argty),*) -> $ret {
                $imp(self, $($argname),*)
            }
            )*
        }

        $(
        #[cfg(feature = "std")]
        $(#[$meta])*
        pub fn $name($($argname:$argty),*) -> $ret {
            impl GlobalRng {
                $(#[$meta])*
                fn $name(&mut self, $($argname:$argty),*) -> $ret {
                    $imp(self, $($argname),*)
                }
            }

            GlobalRng::$name(&mut GlobalRng, $($argname),*)
        }
        )*
    }
}

define_ext! {
    /// Generate a 32-bit floating point number in the specified range.
    fn f32_range(&mut self, range: impl RangeBounds<f32>) -> f32 => float_range::f32;

    /// Generate a 64-bit floating point number in the specified range.
    fn f64_range(&mut self, range: impl RangeBounds<f64>) -> f64 => float_range::f64;

    /// Generate a 32-bit floating point number in the normal distribution with
    /// mean mu and standard deviation sigma.
    #[cfg(any(feature = "std", feature = "libm"))]
    fn f32_normal(&mut self, mu: f32, sigma: f32) -> f32 => float_normal::f32;

    /// Generate a 64-bit floating point number in the normal distribution with
    /// mean mu and standard deviation sigma.
    #[cfg(any(feature = "std", feature = "libm"))]
    fn f64_normal(&mut self, mu: f64, sigma: f64) -> f64 => float_normal::f64;

    /// Generate a 32-bit floating point number in the normal distribution with
    /// mean mu and standard deviation sigma using an approximation algorithm.
    fn f32_normal_approx(&mut self, mu: f32, sigma: f32) -> f32 => float_normal::f32_approx;

    /// Generate a 64-bit floating point number in the normal distribution with
    /// mean mu and standard deviation sigma using an approximation algorithm.
    fn f64_normal_approx(&mut self, mu: f64, sigma: f64) -> f64 => float_normal::f64_approx;
}

mod __private {
    #[doc(hidden)]
    pub trait Sealed {}
    impl Sealed for fastrand::Rng {}
}
